llms/openai/openai.go

```diff
@@ -0,0 +1,295 @@
+	topP             float64 = 1
+	frequencyPenalty float64 = 0
+	presencePenalty  float64 = 0
+	n                int64   = 1
+	bestOf           int64   = 1
+	modelName        string  = "text-davinci-003"
+	batchSize        int64   = 20
+	maxRetries       int     = 3
+)
+
+type OpenAI struct {
+	temperature      float64
+	maxTokens        int64
+	topP             float64
+	frequencyPenalty float64
+	presencePenalty  float64
+	n                int64
+	bestOf           int64
+	logitBias        map[string]interface{}
+	streaming        bool // Streaming Unsupported Right Now
+	modelName        string
+	modelKwargs      map[string]interface{}
+	maxRetries       int
+	batchSize        int64
+	stop             []string
+	timeout          *time.Duration
+	client           *gpt.Gpt
+}
+
+type authorizeTransport struct {
+	apiKey string
+	base   http.RoundTripper
+}
+
+func (t *authorizeTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	fmt.Println("Hitting Transport Override")
+	req.Header.Add("X-Authorization", fmt.Sprintf("Bearer %s", t.apiKey))
+	return http.DefaultTransport.RoundTrip(req)
+}
+
+func New(args ...OpenAIInput) (*OpenAI, error) {
+	if len(args) > 1 {
+		return nil, errors.New("more than one config argument not supported")
+	}
+
+	input := OpenAIInput{}
+	if len(args) > 0 {
+		input = args[0]
+	}
+
+	openai := OpenAI{
+		temperature:      temperature,
+		maxTokens:        maxTokens,
+		topP:             topP,
+		frequencyPenalty: frequencyPenalty,
+		presencePenalty:  presencePenalty,
+		n:                n,
+		bestOf:           bestOf,
+		logitBias:        input.LogitBias,
+		streaming:        input.Streaming,
+		modelName:        modelName,
+		modelKwargs:      input.ModelKwargs,
+		batchSize:        batchSize,
+		stop:             input.Stop,
+		timeout:          input.Timeout,
+		maxRetries:       maxRetries,
+	}
+
+	apiKey := os.Getenv("OPENAI_API_KEY")
+
+	if input.OpenAIApiKey != nil {
+		apiKey = *input.OpenAIApiKey
+	}
+
+	if apiKey == "" {
+		return nil, errors.New("OpenAI API key not found")
+	}
+
+	if input.Temperature != nil {
+		openai.temperature = *input.Temperature
+	}
+
+	if input.MaxTokens != nil {
+		openai.maxTokens = *input.MaxTokens
+	}
+
+	if input.TopP != nil {
+		openai.topP = *input.TopP
+	}
+
+	if input.FrequencyPenalty != nil {
+		openai.frequencyPenalty = *input.FrequencyPenalty
+	}
+
+	if input.PresencePenalty != nil {
+		openai.presencePenalty = *input.PresencePenalty
+	}
+
+	if input.N != nil {
+		openai.n = *input.N
+	}
+
+	if input.BestOf != nil {
+		openai.bestOf = *input.BestOf
+	}
+
+	if input.ModelName != nil {
+		openai.modelName = *input.ModelName
+	}
+
+	if input.BatchSize != nil {
+		openai.batchSize = *input.BatchSize
+	}
+
+	if input.MaxRetries != nil {
+		openai.maxRetries = *input.MaxRetries
+	}
+
+	httpClient := http.Client{Transport: &authorizeTransport{apiKey: apiKey}}
+
+	if openai.timeout != nil {
+		httpClient.Timeout = *openai.timeout
+	}
+
+	client := gpt.New()
```

Comment:
@simplesagar @ndimares

I can't figure out a good way using the SDK to add in the authorization header to each request. Tried to write a quick custom transport to append this header to each request but it doesn't look like the sdk is actually using my custom HTTP client.

Let me know if you've encountered this or know of a more elegant way to attach this authorization header to the SDK.